File: Sagittarius-StarterKits/Java/sagittarius/Sagittarius.java
/**
 * Sagittarius - Java Starter Kit
 * Copyright 2013 WillyG Productions
 * @Authors: William Gaul
 */
package sagittarius;

import java.util.HashMap;
import java.io.PrintStream;

public class Sagittarius {

    private SagittariusLinkClient link;
    private String SagPass;
    private HashMap<String, Module> modules;
    private static PrintStream logStream = System.out;
    private static final String LOG_TAG = "Sagittarius";

    public enum ELogLevel {

        LOG_None, LOG_Error, LOG_Warn, LOG_Info, LOG_Debug;
    }
    private static ELogLevel logLevel = ELogLevel.LOG_Debug;

    public Sagittarius(String AppID, String Pass) {
        this.link = new SagittariusLinkClient(this, AppID);
        this.SagPass = Pass;
        this.modules = new HashMap<String, Module>();
    }
    
    public Action CreateAction(String type) {
        Action a = null;
        if (type.equals("get")) {
            a = new GetAction();
        } else if (type.equals("add")) {
            a = new AddAction();
        } else if (type.equals("mod")) {
            a = new ModAction();
        } else if (type.equals("del")) {
            a = new DelAction();
        } else {
            return a;
        }
        a.initializeRequest(new SagRequest(this));
        return a;
    }

    public void RegisterModule(Module m) {
        modules.put(m.getID(), m);
    }

    public Module GetModule(String ID) {
        Module ret = modules.get(ID);
        if (ret == null) {
            LogWarn("GetModule() - module with ID " + ID + " does not exist");
        }
        return ret;
    }
    
    public void SubmitRequest(SagRequest sr) {
        link.TransmitRequest(sr);
    }

    /**
     * CALLBACK FUNCTIONS
     */
    public void OnResponseReceived(String ModuleID, String ActionID, SagResponse resp) {
        if (ModuleID.equals("builtin")) {
            BuiltInOnResponseReceived(ActionID, resp);
            return;
        }
        GetModule(ModuleID).OnResponseReceived(ActionID, resp);
    }
    
    private void BuiltInOnResponseReceived(String ActionID, SagResponse resp) {
        // @TODO: Make more robust
        if (ActionID.equals("mail")) {
            LogInfo("Mail successfully sent!");
        }
    }

    /**
     * MAIL
     */
    public void SendMail(String receiver, String subject, String message, String sender) {
        SagRequest mail = new SagRequest(this).setDestination("/mail").setModuleInfo("builtin", "mail");
        mail.addURLPair("recv", receiver, false);
        mail.addURLPair("subj", subject, false);
        mail.addURLPair("mesg", message, false);
        if (!sender.equals("")) {
            mail.addURLPair("send", sender, false);
        }
        SubmitRequest(mail);
    }

    /**
     * LOG FUNCTIONS
     */
    public static void SetLogStream(PrintStream ps) {
        logStream = ps;
    }

    public static void SetLogLevel(ELogLevel ll) {
        logLevel = ll;
    }

    private static void log(String msg, String cat, String level) {
        logStream.println("[" + cat + "] " + level + ": " + msg);
    }

    public static void LogError(String msg) {
        LogError(msg, LOG_TAG);
    }

    public static void LogError(String msg, String cat) {
        if (logLevel.ordinal() > 0) {
            log(msg, cat, "ERROR");
        }
    }

    public static void LogWarn(String msg) {
        LogWarn(msg, LOG_TAG);
    }

    public static void LogWarn(String msg, String cat) {
        if (logLevel.ordinal() > 1) {
            log(msg, cat, "WARN");
        }
    }

    public static void LogInfo(String msg) {
        LogInfo(msg, LOG_TAG);
    }

    public static void LogInfo(String msg, String cat) {
        if (logLevel.ordinal() > 2) {
            log(msg, cat, "INFO");
        }
    }

    public static void LogDebug(String msg) {
        LogDebug(msg, LOG_TAG);
    }

    public static void LogDebug(String msg, String cat) {
        if (logLevel.ordinal() > 3) {
            log(msg, cat, "DEBUG");
        }
    }
    
    public String encrypt(String pt) {
        return Encryption.Encrypt(pt, SagPass);
    }

    public String decrypt(String ct) {
        return Encryption.Decrypt(ct, SagPass);
    }
}

File: Sagittarius-StarterKits/JavaScript/sagittarius.js
/**
 * Sagittarius - JavaScript Starter Kit
 * Copyright 2013 WillyG Productions
 * @Authors: William Gaul
 */
function Encryption() {
	// Just to declare the Encrypt namespace
}

Encryption.PadString = function (text, key, pad) {
	var ret = '', seed = 0, i;
	for (i = 0; i < key.length; i++) {
		seed += ((key.charCodeAt(i) & 0xff) * (1 << i));
	}
	seed *= pad;
	for (i = 0; i < text.length; i++) {
		seed = ((seed * 214013 + 2531011) >>> 16) & 0x7fff;
		ret += String.fromCharCode((text.charCodeAt(i) & 0xff) ^ (seed & 0xff));
	}
	return ret;
};

Encryption.GetHexForByte = function (b) {
	var hex = '0123456789ABCDEF', hi = (b >>> 4) & 0x0f, lo = b & 0x0f;
	return hex.charAt(hi) + hex.charAt(lo);
};

Encryption.GetByteForHex = function (s) {
	var hex = '0123456789ABCDEF', hi = hex.indexOf(s.charAt(0)), lo = hex.indexOf(s.charAt(1));
	return ((hi << 4) + lo) & 0xff;
};

Encryption.Encrypt = function (pt, key) {
	var ct = '', pad = Math.floor(Math.random() * 256), temp = Encryption.PadString(pt, key, pad), i;
	for (i = 0; i < temp.length; i++) {
		ct += Encryption.GetHexForByte(temp.charCodeAt(i) & 0xff);
	}
	return "~" + Encryption.GetHexForByte(pad & 0xff) + ct;
};

Encryption.Decrypt = function (ct, key) {
	var temp = '', i;
	for (i = 3; i < ct.length; i += 2) {
		temp += String.fromCharCode(Encryption.GetByteForHex(ct.substring(i, i + 2)));
	}
	return Encryption.PadString(temp, key, Encryption.GetByteForHex(ct.substring(1, 3)));
};

var SagRequest = function (password) {
	var that = {}, data = [];
	that.password = password;

	that.setDestination = function (dest) {
		that.dest = dest;
		return that;
	}

	that.addURLPair = function (key, value, encrypt) {
		if (encrypt !== undefined) {
			value = Encryption.Encrypt(value, that.password);
		}
		data.push({key: key, value: value});
	};
	
	that.setURLPair = function (key, value, encrypt) {
		for (var i = 0; i < urlstring.length; i++) {
			if (urlstring[i].key === key) {
				if (encrypt !== undefined) {
					value = Encryption.Encrypt(value, that.password);
				}
				urlstring[i].value = value;
				return;
			}
		}
	};

	that.GetData = function () {
		var segments = [];
		for (var i = 0; i < urlstring.length; i++) {
			segments[i] = urlstring[i].key + "=" + urlstring[i].value;
		}
		return segments.join('&');
	};

	return that;
}

var Action = function (handler, password) {
	var that = {};
	that.DBTYPE = 'object_type';
	that.DBNAME = 'object_name';
	that.request = SagRequest(password);
	that.request.setDestination(handler);
	return that;
};

var GetAction = function (password) {
	var that = Action('/dbget', password);
	that.request.addURLPair('rlim', 20);
	that.request.addURLPair('roff', 0);
	
	that.AddFilter = function (field, value, encrypt) {
		that.request.addURLPair('f', field + '::' + value, encrypt);
		return that;
	};
	
	that.AddProjection = function (field, encrypt) {
		that.request.addURLPair('p', field + (encrypt !== undefined ? '~' : ''));
		return that;
	};
	
	that.SetLimit = function (rl) {
		that.request.setURLPair('rlim', rl);
		return that;
	};
	
	that.SetOffset = function (ro) {
		that.request.setURLPair('roff', ro);
		return that;
	};
	
	that.Unique = function () {
		that.request.setURLPair('rlim', 1);
		return that;
	};
	
	return that;
};

var AddAction = function (password) {
	var that = Action('/dbadd', password);
	
	that.AddAttribute = function (field, value, encrypt) {
		that.request.addURLPair('a', field + '::' + value, encrypt);
		return that;
	};
	
	return that;
};

var DelAction = function (password) {
	var that = GetAction(password);
	that.handler = '/dbdel';
	that.request.addURLPair('rres', 'false');
	
	that.SetReturnsResults = function () {
		that.request.setURLPair('rres', 'true');
		return that;
	};
	
	return that;
};

var ModAction = function (password) {
	var that = DelAction(password);
	that.handler = '/dbmod';
	
	that.AddModification = function (field, value, encrypt) {
		that.request.addURLPair('m', field + '::' + value, encrypt);
		return that;
	};
	
	return that;
};

function createCORSRequest() {
	var xhr = new XMLHttpRequest();
	if (xhr.hasOwnProperty('withCredentials')) {
		// We're good!
	} else if (typeof XDomainRequest !== 'undefined') {
		xhr = new XDomainRequest();
	} else {
		xhr = null;
	}
	return xhr;
}

function Sagittarius(appid, pass) {
	this.appid = appid;
	this.pass = pass;
}

Sagittarius.prototype.CreateAction = function (type) {
	switch (type) {
		case 'get': return GetAction(this.pass);
		case 'add': return AddAction(this.pass);
		case 'mod': return ModAction(this.pass);
		case 'del': return DelAction(this.pass);
		default: break;
	}
};

Sagittarius.prototype.SubmitAction = function (action, callback) {
	this.TransmitRequest(action.request, callback);
};

Sagittarius.prototype.TransmitRequest = function (sagRequest, callback) {
	var request = createCORSRequest();
	if (!request) {
		return;
	}
	request.open('POST', "http://" + this.appid + ".appspot.com" + sagRequest.dest, true);
	if (window.XDomainRequest) {
		request.onload = function () {
			callback(JSON.parse(request.responseText.replace("<resp>", "").replace("</resp>", "")));
		};
	} else {
		request.setRequestHeader('Content-type', 'application/x-www-form-urlencoded');
		request.onreadystatechange = function (response) {
			if (request.readyState === 4) {
				callback(JSON.parse(request.responseText.replace("<resp>", "").replace("</resp>", "")));
			}
		};
	}
	// Use encodeURI() to encode entire data!!!
	request.send(encodeURI(sagRequest.GetData()));
};

Sagittarius.prototype.SendMail = function (sendObj) {
	var mail = SagRequest(this.pass);
	mail.setDestination('/mail');
	mail.addURLPair('recv', sendObj.receiver);
	mail.addURLPair('subj', sendObj.subject);
	mail.addURLPair('mesg', sendObj.message);
	if (sendObj.hasOwnProperty('sender')) {
		mail.addURLPair('send', sendObj.sender);
	}
	this.TransmitRequest(mail, sendObj.callback);
};